A43-sdk-authorization.md

```diff
@@ -0,0 +1,364 @@
+	  "scoped by services.",
+	"type": "object",
+	"properties": {
+		"name": {
+			"description": "The name of an authorization policy. It is mainly for monitoring and error message generation.",
+			"type": "string",
+		},
+		"deny_rules": {
+			"description": "List of deny rules to match. If a request matches any of the deny rules, then it will be denied. If none of"
+			  "the deny rules matches or there are no deny rules, the allow rules will be evaluated.",
+			"type": "array",
+			"items": {
+				"$ref": "#/definitions/rule",
+			},
+		},
+		"allow_rules": {
+			"description": "List of allow rules to match. The allow rules will only be evaluated after the deny rules. If a request"
+			  "matches any of the allow rules, then it will allowed. If none of the allow rules matches, it will be denied.",
+			"type": "array",
+			"items": {
+				"$ref": "#/definitions/rule",
+			},
+		}
+	},
+	"required": ["name", "allow_rules"]
+}
+```
+
+gRPC SDK authorization policy has a list of allow rules and a list of deny rules. 
+The following sequence is followed to make an authorization decision -
+1. Check for a match in deny rules, if a match is found, the request will be denied. 
+   If no match is found in deny rules, or there are no deny rules execute next step.
+2. Check for a match in allow rules, if a match is found, the request will be allowed. 
+   Note that allow rules is a required field here, if not present, policy is invalid.
+3. If no match is found in allow rules, we deny the request.
+
+In the following policy example
+- Peer identity from ["spiffe://foo.com/sa/admin1", "spiffe://foo.com/sa/admin2"] is 
+  authorized to access any RPC methods in pkg.service
+- Any authenticated user is allowed to access "foo" and "bar" RPC methods.
+- Nobody can access "secret" RPC method, not even the admins as deny rules are evaluated 
+  first.
+
+```json
+{
+	"name": "example-policy",
+	"allow_rules": [
+		{
+			"name": "admin-access",
+			"source": {
+				"principals": [
+					"spiffe://foo.com/sa/admin1",
+					"spiffe://foo.com/sa/admin2"
+				]
+			},
+			"request": {
+				"paths": ["/pkg.service/*"]
+			}
+		},
+		{
+			"name": "dev-access",
+			"source": {
+				"principals": [
+					"*"
+				]
+			},
+			"request": {
+				"paths": [
+					"/pkg.service/foo",
+					"/pkg.service/bar"
+				]
+			}
+		}
+	],
+	"deny_rules": [
+		{
+			"name": "deny-access",
+			"request": {
+				"paths": [
+					"*/secret"
+				]
+			}
+		}
+	]
+}
+```
+
+As mentioned previously, internally gRPC Authorization engine consumes Envoy RBAC 
+proto. The user provided JSON SDK authorization policy will be translated to Envoy 
+RBAC proto and provided to Engines for evaluation. Note that SDK authorization policy 
+is a subset of Envoy RBAC, and it does not support all the fields that are present in 
+Envoy RBAC.
+
+### API
+
+gRPC will support both static initialization and dynamically reloading the policy
```

Comment:
There are a couple of interop issues with xDS.

1) If server is an xDS server and receives authz policies from the xDS control plane, how will those policies and the SDK supplied authz policies interact? If this is an error, how will it be caught?

2) If the server also uses XdsServerCredentials and it tells the Server to use only TLS or plaintext, what happens to SDK supplied authz policies?

Reply:
They are all great questions!

Reply:
I think if the server is xDS server or uses XdsServerCredentials, it should **only** try the authz policy delivered by xDS control plane (if provided) because there is no guarantee that xDS authz and SDK authz will not diverge in semantics in future. @ejona86, @markdroth, @dfawley, @sanjaypujare, WDYT?

Reply:
Maybe we could have SDK authorization in xDS enabled servers. In which case, ex. C++, both the filters (xDS and non-xDS) may be executed. Since they are separate filters, we have the flexibility to arrange SDK filter to be higher in stack relative to xDS.

Reply:
> ...In which case, ex. C++, both the filters (xDS and non-xDS) may be executed....

I am okay with this provided the semantics are well defined and everyone agrees with them.

Another idea I had was something similar to "fallback credentials" of XdsCredentials. We can define our SDK API in such a way so that it is possible for the user to say that use the xDS provided authz policy if present otherwise use this policy (the one I am supplying thru the API).

Reply:
I am still not convinced we should allow the use of SDK authorization in xDS servers because it will give users an impression that SDK authz will be always compatible with xDS authz for which we do not have the guarantee. If we want to support the idea of "fallback authz" for xDS servers, the authz check should also follow the xDS authz semantics.

Reply:
I don't see any problem here. xDS will run its policies, and SDK will run its policies. This seems no different than the "multiple SDK interceptors" case which we recently agreed was fine. We aren't encouraging it, but if someone does it things are okay.

Note Java/Go have no technical means to prevent mixing this authz with xDS.

Reply:
Thank you for the discussion. I think we have agreement that SDK authorization can be enabled in xDS enabled servers, allowing both paths to co-exist. I added a statement in the rationale section.